### PR TITLE
Enable --expire-all-object-versions

### PR DESCRIPTION
**Pull request by shtripat**
## Community Contribution License
All community contributions in this pull request are licensed to the project maintainers
under the terms of the [Apache 2 license](https://www.apache.org/licenses/LICENSE-2.0). 
By creating this pull request I represent that I have the right to license the 
contributions to the project maintainers under the Apache 2 license.

## Description
This PR introduces a new flag `--expire-all-object-versions` to the command `mc ilm rule add` to enabled option to create ILM rule to remove all versions of expired objects.

## Motivation and Context
Allow setting ILM rule for removing all versions of expired objects from command line.

## How to test this PR?
Follow below steps to verify the changes

Step-1: Set your system date to few days older (~3days)

Step-2: Create a MinIO instance, create a bucket and enable versioning for the bucket

Step-3: Set the bucket public using `mc anonymous set public ALIAS/BUCKET`

Step-4: Set the MinIO scanner speed to fastest using `mc admin config set ALIAS scanner speed=fastest`

Step-5: Load say 100 objects to the bucket using curl command with older mtime value as below
```
curl -X PUT http://127.0.0.1:9000/BUCKET/OBJNAME -d @PATH-TO-OBJ-SRC-FILE -H "X-Minio-Source-Mtime: 2023-03-12T16:19:03.115209866-08:00"
```

Step-6: Now delete these objects using `mc rm` which would create DEL markers for all the objects. Eventually two versions for all objects

Step-7: Set the date back to current for the system

Step-8: Add bewlo ILM rules for bucket
`mc ilm rule add ALIAS/BUCKET --expire-all-object-versions --expire-days 1`

Step-9: After few seconds all versions of all objects should be removed and `mc ls ALIAS/BUCKET --versions` should not return any values.

Step:10: Also verify prometheus metrics values `minio_s3_requests_total{api="putobject",server="127.0.0.1:9000"}` and `minio_node_ilm_action_count_delete_all_versions_action{server="127.0.0.1:9000"}`. They should return 100 and 200 respectively as count values.

Note: needs: https://github.com/minio/minio-go/pull/1927 and https://github.com/minio/minio/pull/18954/

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fmc%2fpull%2fNNNNN)

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -131,3 +131,5 @@ require (
 	google.golang.org/protobuf v1.31.0 // indirect
 	gopkg.in/ini.v1 v1.67.0 // indirect
 )
+
+replace github.com/minio/minio-go/v7 v7.0.66 => github.com/shtripat/minio-go/v7 v7.0.0-20240202175226-e4e2f9e24339
```
Comment (harshavardhana) [inline go.mod:135]: please update the replace flag

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: go.mod (hunk 1 above)
Comment (shtripat) [reply go.mod]: done

File: cmd/ilm-rule-add.go [hunk 2]
```diff
@@ -182,6 +182,10 @@ var ilmAddFlags = []cli.Flag{
 		Name:  "noncurrent-transition-tier",
 		Usage: "remote tier name to transition",
 	},
+	cli.BoolFlag{
+		Name:  "expired-object-all-versions",
```
Comment (vadmeste) [inline cmd/ilm-rule-add.go:186]: maybe 'expire-all-object-versions' ?

File: cmd/ilm/options.go [hunk 3]
```diff
@@ -429,6 +435,12 @@ func ApplyRuleFields(dest *lifecycle.Rule, opts LifecycleOptions) *probe.Error {
 		dest.Expiration.DeleteMarker = lifecycle.ExpireDeleteMarker(*opts.ExpiredObjectDeleteMarker)
 		dest.Expiration.Days = 0
 		dest.Expiration.Date = lifecycle.ExpirationDate{}
+		// safe to check for delete here as delete marker would at least always come as false
+		if opts.ExpiredObjectAllversions != nil {
```
Comment (vadmeste) [inline cmd/ilm/options.go:439]: why this 'if' is inside 	'} else if opts.ExpiredObjectDeleteMarker != nil {' ?

File: cmd/ilm/options.go [hunk 4]
```diff
@@ -429,6 +435,12 @@ func ApplyRuleFields(dest *lifecycle.Rule, opts LifecycleOptions) *probe.Error {
...
+			dest.Expiration.DeleteAll = lifecycle.ExpirationBoolean(*opts.ExpiredObjectAllversions)
+			dest.Expiration.Days = 0
+			dest.Expiration.Date = lifecycle.ExpirationDate{}
```
Comment (vadmeste) [inline cmd/ilm/options.go:442]: Why you have to reset both fields here ? this is different to expired object delete marker

**Review by vadmeste — CHANGES_REQUESTED** [review]
Few questions

**Review by harshavardhana — APPROVED** [review]

**Review by vadmeste — APPROVED** [review]
LGTM

**Review by prakashsvmx — APPROVED** [review]
Tested . LGTM :+1:

**Review by harshavardhana — APPROVED** [review]

**Pull request merged**